scripts/solve_network.py

```diff
@@ -854,6 +854,43 @@ def add_existing(n):
             n.generators.loc[tech_index, tech] = existing_res
 
 
+def add_lossy_bidirectional_link_constraints(n: pypsa.components.Network) -> None:
+    """
+    Ensures that the two links simulating a bidirectional_link are extended the same amount.
+    """
+
+    if not n.links.p_nom_extendable.any() or "reversed" not in n.links.columns:
+        return
+
+    n.links["reversed"] = n.links.reversed.fillna(0).astype(bool)
+    carriers = n.links.loc[n.links.reversed, "carrier"].unique()  # noqa: F841

```

Comment:
Short answer: carriers is used in the lines just below, but the interpreter doesn't realize it, because its used in with .query(). Thats why noqa: F841 is used.

I'll write a long answer tomorrow.

Reply:
Ahh, ok! Thanks a lot for investigating this! Agree with the proposal on replacing here `query` with `loc`

Regarding the overall code refactoring, totally agree that there is a room for improvement and you are very welcome to contribute there. Frankly speaking, I'm not ready to confirm that `query` is less readable than `loc` for all the cases. But you may be right that it could be over-used for sure, and happy to discuss you propositions!